toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -87,7 +113,7 @@ class DiagnosticConsumer {
   virtual ~DiagnosticConsumer() = default;
 
   // Handle a diagnostic.
-  virtual auto HandleDiagnostic(const Diagnostic& diagnostic) -> void = 0;
+  virtual auto HandleDiagnostic(Diagnostic diagnostic) -> void = 0;
```

Comment:
Why this change? Seems somewhat wasteful to copy (or even move) the diagnostic objects here... They're going to end up quite large with `std::function` objects and small vectors, etc.

I could also imagine wanting to support multiple consumers directly (instead of the chaining done by wrapping consumers below), but that's maybe not relevant yet.

Reply:
The consequence of using a const& here is that the SortingDiagnosticConsumer needs to copy, since it can't move. Do you prefer the copy instead of moves?

Reply:
It's an issue on this PR because the implicit copy no longer works -- pointers can't be copied. So I can either implement a custom copy that duplicates the pointers, or do moves. Either way I need to implement something here. I don't think "leave this API" is an option.

Reply:
Ah, the fact that we now have `unique_ptr` (and not a more `Box` like construct) is what eluded, me, thanks!

I'm happy moving for now, even if it makes sense to revisit it later.

Reply:
Per discussion, will send a followup that:

- Documents why we move more explicitly.
- Switch to a wrapper class to try getting rid of unique_ptr (but still not copying).